docs/common/LOCAL_DEV_SERVERS.md

```diff
@@ -117,6 +119,47 @@ tail -f packages/shell/local-dev-shell.log
 tail -f packages/toolshed/local-dev-toolshed.log
 ```
 
+### OAuth Returns HTML Instead of JSON
+
+**Error:** `SyntaxError: Unexpected token '<', "<!DOCTYPE "... is not valid JSON`
+
+This happens when OAuth or API calls hit port 5173 (frontend) instead of port 8000 (backend). The frontend returns HTML, which fails JSON parsing.
+
+| Port | Server | Returns |
+|------|--------|---------|
+| **5173** | Frontend (Shell) | HTML |
+| **8000** | Backend (Toolshed) | JSON |
+
+**Fix:** Ensure API calls use port 8000:
+- Deployment: `--api-url http://localhost:8000`
+- In patterns: Check `getRecipeEnvironment().apiUrl`
+- Browser testing: Navigate to `http://localhost:5173` for UI, but API calls should target 8000
+
+### UI Component Changes Not Rebuilding
+
+When modifying UI components in `packages/ui/src/v2/components/`, the shell dev server doesn't automatically rebuild because it only watches `packages/shell/src/`.
+
+**Symptom:** You modify a UI component, redeploy a charm, but your changes don't appear.
+
+**Fix:** Touch a file in shell/src to trigger a full rebuild:
+
+```bash
+# Touch shell source to trigger rebuild
+touch packages/shell/src/index.ts
+
+# Watch the log for rebuild confirmation
+tail -f packages/shell/local-dev-shell.log
+# Should see: "🔨 Building..." then "✓ Built..."
+```
+
+**Why this works:** The shell's esbuild config bundles all dependencies (including `@commontools/ui`). When any watched file changes, esbuild rebuilds the entire bundle with the latest UI component code.
+
+**Verify your changes are included:**
+```bash
+# Search for a unique string from your changes
+grep "your-unique-string" packages/shell/dist/scripts/index.js
+```
+
```

Comment:
Invert this guidance and it becomes one sentence: when editing `ct-` components, restart local dev server to ensure the updated code is running.